cmd/sourced/cmd/web.go

```diff
@@ -32,54 +33,177 @@ func init() {
+// 2vudhlzztdlbg_ghsync_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   ./init.sh                        Up                      0.0.0.0:3306->3306/tcp
+// 2vudhlzztdlbg_gitbase_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /bin/dumb-init -- /bin/sh  ...   Up
+// 2vudhlzztdlbg_gitcollector_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /home/entrypoint.sh start- ...   Up                      0.0.0.0:10000->10000/tcp, 0.0.0.0:4040->4040/tcp, 8888/tcp
+// 2vudhlzztdlbg_gsc_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh postgres    Up                      0.0.0.0:5433->5432/tcp
+// 2vudhlzztdlbg_metadatadb_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh postgres    Up                      0.0.0.0:5432->5432/tcp
+// 2vudhlzztdlbg_postgres_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   docker-entrypoint.sh redis ...   Up                      0.0.0.0:6379->6379/tcp
+// 2vudhlzztdlbg_redis_1
+// srcd-l1vzzxjzl3nln2vudhlzztdlbi9qcm9qzwn0cy8uz28td29ya3nwywnll3nyyy9naxrodwiuy29tl3nln   /entrypoint.sh                   Up (health: starting)   0.0.0.0:8088->8088/tcp
+// 2vudhlzztdlbg_sourced-ui_1
+func runMonitor(ch chan<- error) {
+	runMonitorService := func(service string, ch chan<- error) {
+		for {
+			var stdout bytes.Buffer
+			if err := compose.RunWithIO(context.Background(),
+				os.Stdin, &stdout, nil, "ps", service); err != nil {
+				ch <- errors.Wrapf(err, "cannot get status service %s", service)
+				return
+			}
+
+			matches := stateExtractor.FindAllStringSubmatch(
+				normalizeNewLine(strings.TrimSpace(stdout.String())), -1)
+			for _, match := range matches {
+				state := match[1]
+
+				if strings.HasPrefix(state, "Exit") {
+					if service != "ghsync" && service != "gitcollector" {
+						ch <- fmt.Errorf("service '%s' is in state '%s'", service, state)
+						return
+					}
+
+					returnCode := state[len("Exit "):len(state)]
+					if returnCode != "0" {
+						ch <- fmt.Errorf("service '%s' exited with return code: %s",
+							service, returnCode)
+						return
+					}
+
+					continue
+				}
+
+				if state != "Up" {
+					ch <- fmt.Errorf("service '%s' is in state '%s'", service, state)
+					return
+				}
+
+			}
+
+			time.Sleep(1 * time.Second)
+		}
 	}
 
-	return nil
+	go func() {
+		var servicesBuf bytes.Buffer
+		if err := compose.RunWithIO(context.Background(),
+			os.Stdin, &servicesBuf, nil, "config", "--services"); err != nil {
+			ch <- errors.Wrap(err, "cannot get list of services")
+			return
+		}
+
+		services := strings.Split(normalizeNewLine(
+			strings.TrimSpace(servicesBuf.String())), "\n")
+
+		for _, service := range services {
+			go runMonitorService(service, ch)
```

Comment:
what is the advantage of using goroutines here?

Reply:
The only reason is to be more reactive to state changes in containers. The alternative that I thought is to remove the `while true` and the `sleep` from the `runMonitorService` function and to write this for loop as:
```
for {
    for _, service := range services {
        runMonitorService(service, ch)
    }
}
```

The problem is that the call to `docker-compose ps <service>` takes an amount of time near to ~1 second (at least in my setup). So with the non-gorountine solution, the container of a service is checked once every ~`n`-seconds where `n` is the number of services. With the gorountine solution empirically each service each checked much faster. With the `sleep` of 1 second as now, every service is checked every ~4 seconds. It's to be noted that this is based only on empirical observations and that I'm pretty sure also for the gorountine solution the amount of time is a function of `n` too.
Long story short, with our current number of services docker answers faster to all our requests if they're made concurrently instead of serially. This may change if in future we add more services.

Reply:
So if we're happy with a less reactive solution in favor of easier code, I'm also ok with that. I don't think that a user would get mad for waiting a couple of seconds more.

Reply:
I don't have a strong opinion about one or another approach.
If I was writing it from scratch I would choose simpler code over 2 seconds delay of error reporting (taking into account that the user sees the loader at this time). But if you prefer to keep it as it is now, it's fine.

Reply:
I'm ok with both really 😄. Let's see also what @dpordomingo and @carlosms says. 👍